// thirdweb/dist/esm/wallets/in-app/native/helpers/wallet/sharded-wallet.js
import { fetchUserDetails } from "../api/fetchers.js";
import { postAuth } from "../auth/middleware.js";
import { getWalletUserDetails } from "../storage/local.js";
import { getExistingUserAccount } from "./retrieval.js";
export class ShardedWallet {
    constructor(args) {
        Object.defineProperty(this, "client", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "storage", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.client = args.client;
        this.storage = args.storage;
    }
    async postWalletSetUp(authResult) {
        await postAuth({
            storedToken: authResult.storedToken,
            client: this.client,
            storage: this.storage,
            encryptionKey: authResult.encryptionKey,
        });
    }
    async getUserWalletStatus() {
        const localData = await getWalletUserDetails(this.client.clientId);
        const userStatus = await fetchUserDetails({
            client: this.client,
            email: localData?.email,
            storage: this.storage,
        });
        if (userStatus.status === "Logged In, Wallet Initialized") {
            return {
                status: userStatus.status,
                authDetails: userStatus.storedToken.authDetails,
                walletAddress: userStatus.walletAddress,
                account: await this.getAccount(),
            };
        }
        if (userStatus.status === "Logged In, New Device") {
            return {
                status: "Logged In, New Device",
                authDetails: userStatus.storedToken.authDetails,
                walletAddress: userStatus.walletAddress,
            };
        }
        if (userStatus.status === "Logged In, Wallet Uninitialized") {
            return {
                status: "Logged In, Wallet Uninitialized",
                authDetails: userStatus.storedToken.authDetails,
            };
        }
        // Logged out
        return { status: "Logged Out" };
    }
    getAccount() {
        return getExistingUserAccount({
            client: this.client,
            storage: this.storage,
        });
    }
}
