// thirdweb/dist/esm/wallets/in-app/web/lib/actions/sign-transaction.enclave.js
import { getThirdwebBaseUrl } from "../../../../../utils/domains.js";
import { getClientFetch } from "../../../../../utils/fetch.js";
import { stringify } from "../../../../../utils/json.js";
export async function signTransaction({ client, ecosystem, payload, storage, }) {
    console.log("payload", payload);
    const clientFetch = getClientFetch(client, ecosystem);
    const authToken = await storage.getAuthCookie();
    const response = await clientFetch(`${getThirdwebBaseUrl("inAppWallet")}/api/v1/enclave-wallet/sign-transaction`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            "x-thirdweb-client-id": client.clientId,
            Authorization: `Bearer embedded-wallet-token:${authToken}`,
        },
        body: stringify({
            transactionPayload: payload,
        }),
    });
    if (!response.ok) {
        throw new Error("Failed to sign transaction");
    }
    const signedTransaction = (await response.json());
    return signedTransaction.signature;
}
