// thirdweb/dist/esm/extensions/pack/__generated__/IPack/read/getTokenOfBundle.js
import { readContract } from "../../../../../transaction/read-contract.js";
import { encodeAbiParameters } from "../../../../../utils/abi/encodeAbiParameters.js";
import { decodeAbiParameters } from "viem";
import { detectMethod } from "../../../../../utils/bytecode/detectExtension.js";
export const FN_SELECTOR = "0x1da799c9";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "_bundleId",
    },
    {
        type: "uint256",
        name: "index",
    },
];
const FN_OUTPUTS = [
    {
        type: "tuple",
        components: [
            {
                type: "address",
                name: "assetContract",
            },
            {
                type: "uint8",
                name: "tokenType",
            },
            {
                type: "uint256",
                name: "tokenId",
            },
            {
                type: "uint256",
                name: "totalAmount",
            },
        ],
    },
];
/**
 * Checks if the `getTokenOfBundle` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getTokenOfBundle` method is supported.
 * @extension PACK
 * @example
 * ```ts
 * import { isGetTokenOfBundleSupported } from "thirdweb/extensions/pack";
 * const supported = isGetTokenOfBundleSupported(["0x..."]);
 * ```
 */
export function isGetTokenOfBundleSupported(availableSelectors) {
    return detectMethod({
        availableSelectors,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getTokenOfBundle" function.
 * @param options - The options for the getTokenOfBundle function.
 * @returns The encoded ABI parameters.
 * @extension PACK
 * @example
 * ```ts
 * import { encodeGetTokenOfBundleParams } from "thirdweb/extensions/pack";
 * const result = encodeGetTokenOfBundleParams({
 *  bundleId: ...,
 *  index: ...,
 * });
 * ```
 */
export function encodeGetTokenOfBundleParams(options) {
    return encodeAbiParameters(FN_INPUTS, [options.bundleId, options.index]);
}
/**
 * Encodes the "getTokenOfBundle" function into a Hex string with its parameters.
 * @param options - The options for the getTokenOfBundle function.
 * @returns The encoded hexadecimal string.
 * @extension PACK
 * @example
 * ```ts
 * import { encodeGetTokenOfBundle } from "thirdweb/extensions/pack";
 * const result = encodeGetTokenOfBundle({
 *  bundleId: ...,
 *  index: ...,
 * });
 * ```
 */
export function encodeGetTokenOfBundle(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (FN_SELECTOR +
        encodeGetTokenOfBundleParams(options).slice(2));
}
/**
 * Decodes the result of the getTokenOfBundle function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension PACK
 * @example
 * ```ts
 * import { decodeGetTokenOfBundleResult } from "thirdweb/extensions/pack";
 * const result = decodeGetTokenOfBundleResultResult("...");
 * ```
 */
export function decodeGetTokenOfBundleResult(result) {
    return decodeAbiParameters(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getTokenOfBundle" function on the contract.
 * @param options - The options for the getTokenOfBundle function.
 * @returns The parsed result of the function call.
 * @extension PACK
 * @example
 * ```ts
 * import { getTokenOfBundle } from "thirdweb/extensions/pack";
 *
 * const result = await getTokenOfBundle({
 *  contract,
 *  bundleId: ...,
 *  index: ...,
 * });
 *
 * ```
 */
export async function getTokenOfBundle(options) {
    return readContract({
        contract: options.contract,
        method: [FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.bundleId, options.index],
    });
}
