// thirdweb/dist/cjs/extensions/pack/__generated__/IPack/read/getPackContents.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetPackContentsSupported = isGetPackContentsSupported;
exports.encodeGetPackContentsParams = encodeGetPackContentsParams;
exports.encodeGetPackContents = encodeGetPackContents;
exports.decodeGetPackContentsResult = decodeGetPackContentsResult;
exports.getPackContents = getPackContents;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x8d4c446a";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "_packId",
    },
];
const FN_OUTPUTS = [
    {
        type: "tuple[]",
        name: "contents",
        components: [
            {
                type: "address",
                name: "assetContract",
            },
            {
                type: "uint8",
                name: "tokenType",
            },
            {
                type: "uint256",
                name: "tokenId",
            },
            {
                type: "uint256",
                name: "totalAmount",
            },
        ],
    },
    {
        type: "uint256[]",
        name: "perUnitAmounts",
    },
];
/**
 * Checks if the `getPackContents` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getPackContents` method is supported.
 * @extension PACK
 * @example
 * ```ts
 * import { isGetPackContentsSupported } from "thirdweb/extensions/pack";
 * const supported = isGetPackContentsSupported(["0x..."]);
 * ```
 */
function isGetPackContentsSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getPackContents" function.
 * @param options - The options for the getPackContents function.
 * @returns The encoded ABI parameters.
 * @extension PACK
 * @example
 * ```ts
 * import { encodeGetPackContentsParams } from "thirdweb/extensions/pack";
 * const result = encodeGetPackContentsParams({
 *  packId: ...,
 * });
 * ```
 */
function encodeGetPackContentsParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.packId]);
}
/**
 * Encodes the "getPackContents" function into a Hex string with its parameters.
 * @param options - The options for the getPackContents function.
 * @returns The encoded hexadecimal string.
 * @extension PACK
 * @example
 * ```ts
 * import { encodeGetPackContents } from "thirdweb/extensions/pack";
 * const result = encodeGetPackContents({
 *  packId: ...,
 * });
 * ```
 */
function encodeGetPackContents(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetPackContentsParams(options).slice(2));
}
/**
 * Decodes the result of the getPackContents function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension PACK
 * @example
 * ```ts
 * import { decodeGetPackContentsResult } from "thirdweb/extensions/pack";
 * const result = decodeGetPackContentsResultResult("...");
 * ```
 */
function decodeGetPackContentsResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result);
}
/**
 * Calls the "getPackContents" function on the contract.
 * @param options - The options for the getPackContents function.
 * @returns The parsed result of the function call.
 * @extension PACK
 * @example
 * ```ts
 * import { getPackContents } from "thirdweb/extensions/pack";
 *
 * const result = await getPackContents({
 *  contract,
 *  packId: ...,
 * });
 *
 * ```
 */
async function getPackContents(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.packId],
    });
}
