// thirdweb/dist/cjs/extensions/pack/__generated__/IPack/read/getTokenCountOfBundle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetTokenCountOfBundleSupported = isGetTokenCountOfBundleSupported;
exports.encodeGetTokenCountOfBundleParams = encodeGetTokenCountOfBundleParams;
exports.encodeGetTokenCountOfBundle = encodeGetTokenCountOfBundle;
exports.decodeGetTokenCountOfBundleResult = decodeGetTokenCountOfBundleResult;
exports.getTokenCountOfBundle = getTokenCountOfBundle;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xd0d2fe25";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "_bundleId",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
    },
];
/**
 * Checks if the `getTokenCountOfBundle` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getTokenCountOfBundle` method is supported.
 * @extension PACK
 * @example
 * ```ts
 * import { isGetTokenCountOfBundleSupported } from "thirdweb/extensions/pack";
 * const supported = isGetTokenCountOfBundleSupported(["0x..."]);
 * ```
 */
function isGetTokenCountOfBundleSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getTokenCountOfBundle" function.
 * @param options - The options for the getTokenCountOfBundle function.
 * @returns The encoded ABI parameters.
 * @extension PACK
 * @example
 * ```ts
 * import { encodeGetTokenCountOfBundleParams } from "thirdweb/extensions/pack";
 * const result = encodeGetTokenCountOfBundleParams({
 *  bundleId: ...,
 * });
 * ```
 */
function encodeGetTokenCountOfBundleParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.bundleId]);
}
/**
 * Encodes the "getTokenCountOfBundle" function into a Hex string with its parameters.
 * @param options - The options for the getTokenCountOfBundle function.
 * @returns The encoded hexadecimal string.
 * @extension PACK
 * @example
 * ```ts
 * import { encodeGetTokenCountOfBundle } from "thirdweb/extensions/pack";
 * const result = encodeGetTokenCountOfBundle({
 *  bundleId: ...,
 * });
 * ```
 */
function encodeGetTokenCountOfBundle(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetTokenCountOfBundleParams(options).slice(2));
}
/**
 * Decodes the result of the getTokenCountOfBundle function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension PACK
 * @example
 * ```ts
 * import { decodeGetTokenCountOfBundleResult } from "thirdweb/extensions/pack";
 * const result = decodeGetTokenCountOfBundleResultResult("...");
 * ```
 */
function decodeGetTokenCountOfBundleResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getTokenCountOfBundle" function on the contract.
 * @param options - The options for the getTokenCountOfBundle function.
 * @returns The parsed result of the function call.
 * @extension PACK
 * @example
 * ```ts
 * import { getTokenCountOfBundle } from "thirdweb/extensions/pack";
 *
 * const result = await getTokenCountOfBundle({
 *  contract,
 *  bundleId: ...,
 * });
 *
 * ```
 */
async function getTokenCountOfBundle(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.bundleId],
    });
}
