// thirdweb/dist/cjs/extensions/pack/__generated__/IPack/read/getUriOfBundle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetUriOfBundleSupported = isGetUriOfBundleSupported;
exports.encodeGetUriOfBundleParams = encodeGetUriOfBundleParams;
exports.encodeGetUriOfBundle = encodeGetUriOfBundle;
exports.decodeGetUriOfBundleResult = decodeGetUriOfBundleResult;
exports.getUriOfBundle = getUriOfBundle;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x61195e94";
const FN_INPUTS = [
    {
        type: "uint256",
        name: "_bundleId",
    },
];
const FN_OUTPUTS = [
    {
        type: "string",
    },
];
/**
 * Checks if the `getUriOfBundle` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getUriOfBundle` method is supported.
 * @extension PACK
 * @example
 * ```ts
 * import { isGetUriOfBundleSupported } from "thirdweb/extensions/pack";
 * const supported = isGetUriOfBundleSupported(["0x..."]);
 * ```
 */
function isGetUriOfBundleSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getUriOfBundle" function.
 * @param options - The options for the getUriOfBundle function.
 * @returns The encoded ABI parameters.
 * @extension PACK
 * @example
 * ```ts
 * import { encodeGetUriOfBundleParams } from "thirdweb/extensions/pack";
 * const result = encodeGetUriOfBundleParams({
 *  bundleId: ...,
 * });
 * ```
 */
function encodeGetUriOfBundleParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.bundleId]);
}
/**
 * Encodes the "getUriOfBundle" function into a Hex string with its parameters.
 * @param options - The options for the getUriOfBundle function.
 * @returns The encoded hexadecimal string.
 * @extension PACK
 * @example
 * ```ts
 * import { encodeGetUriOfBundle } from "thirdweb/extensions/pack";
 * const result = encodeGetUriOfBundle({
 *  bundleId: ...,
 * });
 * ```
 */
function encodeGetUriOfBundle(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetUriOfBundleParams(options).slice(2));
}
/**
 * Decodes the result of the getUriOfBundle function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension PACK
 * @example
 * ```ts
 * import { decodeGetUriOfBundleResult } from "thirdweb/extensions/pack";
 * const result = decodeGetUriOfBundleResultResult("...");
 * ```
 */
function decodeGetUriOfBundleResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getUriOfBundle" function on the contract.
 * @param options - The options for the getUriOfBundle function.
 * @returns The parsed result of the function call.
 * @extension PACK
 * @example
 * ```ts
 * import { getUriOfBundle } from "thirdweb/extensions/pack";
 *
 * const result = await getUriOfBundle({
 *  contract,
 *  bundleId: ...,
 * });
 *
 * ```
 */
async function getUriOfBundle(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.bundleId],
    });
}
