// thirdweb/dist/cjs/extensions/unstoppable-domains/__generated__/UnstoppableDomains/read/getMany.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isGetManySupported = isGetManySupported;
exports.encodeGetManyParams = encodeGetManyParams;
exports.encodeGetMany = encodeGetMany;
exports.decodeGetManyResult = decodeGetManyResult;
exports.getMany = getMany;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x1bd8cc1a";
const FN_INPUTS = [
    {
        type: "string[]",
        name: "keys",
    },
    {
        type: "uint256",
        name: "tokenId",
    },
];
const FN_OUTPUTS = [
    {
        type: "string[]",
        name: "values",
    },
];
/**
 * Checks if the `getMany` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `getMany` method is supported.
 * @extension UNSTOPPABLE-DOMAINS
 * @example
 * ```ts
 * import { isGetManySupported } from "thirdweb/extensions/unstoppable-domains";
 * const supported = isGetManySupported(["0x..."]);
 * ```
 */
function isGetManySupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "getMany" function.
 * @param options - The options for the getMany function.
 * @returns The encoded ABI parameters.
 * @extension UNSTOPPABLE-DOMAINS
 * @example
 * ```ts
 * import { encodeGetManyParams } from "thirdweb/extensions/unstoppable-domains";
 * const result = encodeGetManyParams({
 *  keys: ...,
 *  tokenId: ...,
 * });
 * ```
 */
function encodeGetManyParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.keys, options.tokenId]);
}
/**
 * Encodes the "getMany" function into a Hex string with its parameters.
 * @param options - The options for the getMany function.
 * @returns The encoded hexadecimal string.
 * @extension UNSTOPPABLE-DOMAINS
 * @example
 * ```ts
 * import { encodeGetMany } from "thirdweb/extensions/unstoppable-domains";
 * const result = encodeGetMany({
 *  keys: ...,
 *  tokenId: ...,
 * });
 * ```
 */
function encodeGetMany(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeGetManyParams(options).slice(2));
}
/**
 * Decodes the result of the getMany function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension UNSTOPPABLE-DOMAINS
 * @example
 * ```ts
 * import { decodeGetManyResult } from "thirdweb/extensions/unstoppable-domains";
 * const result = decodeGetManyResultResult("...");
 * ```
 */
function decodeGetManyResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "getMany" function on the contract.
 * @param options - The options for the getMany function.
 * @returns The parsed result of the function call.
 * @extension UNSTOPPABLE-DOMAINS
 * @example
 * ```ts
 * import { getMany } from "thirdweb/extensions/unstoppable-domains";
 *
 * const result = await getMany({
 *  contract,
 *  keys: ...,
 *  tokenId: ...,
 * });
 *
 * ```
 */
async function getMany(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.keys, options.tokenId],
    });
}
