// thirdweb/dist/cjs/extensions/unstoppable-domains/__generated__/UnstoppableDomains/read/namehash.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isNamehashSupported = isNamehashSupported;
exports.encodeNamehashParams = encodeNamehashParams;
exports.encodeNamehash = encodeNamehash;
exports.decodeNamehashResult = decodeNamehashResult;
exports.namehash = namehash;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0x276fabb1";
const FN_INPUTS = [
    {
        type: "string[]",
        name: "labels",
    },
];
const FN_OUTPUTS = [
    {
        type: "uint256",
        name: "hash",
    },
];
/**
 * Checks if the `namehash` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `namehash` method is supported.
 * @extension UNSTOPPABLE-DOMAINS
 * @example
 * ```ts
 * import { isNamehashSupported } from "thirdweb/extensions/unstoppable-domains";
 * const supported = isNamehashSupported(["0x..."]);
 * ```
 */
function isNamehashSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "namehash" function.
 * @param options - The options for the namehash function.
 * @returns The encoded ABI parameters.
 * @extension UNSTOPPABLE-DOMAINS
 * @example
 * ```ts
 * import { encodeNamehashParams } from "thirdweb/extensions/unstoppable-domains";
 * const result = encodeNamehashParams({
 *  labels: ...,
 * });
 * ```
 */
function encodeNamehashParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.labels]);
}
/**
 * Encodes the "namehash" function into a Hex string with its parameters.
 * @param options - The options for the namehash function.
 * @returns The encoded hexadecimal string.
 * @extension UNSTOPPABLE-DOMAINS
 * @example
 * ```ts
 * import { encodeNamehash } from "thirdweb/extensions/unstoppable-domains";
 * const result = encodeNamehash({
 *  labels: ...,
 * });
 * ```
 */
function encodeNamehash(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeNamehashParams(options).slice(2));
}
/**
 * Decodes the result of the namehash function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension UNSTOPPABLE-DOMAINS
 * @example
 * ```ts
 * import { decodeNamehashResult } from "thirdweb/extensions/unstoppable-domains";
 * const result = decodeNamehashResultResult("...");
 * ```
 */
function decodeNamehashResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "namehash" function on the contract.
 * @param options - The options for the namehash function.
 * @returns The parsed result of the function call.
 * @extension UNSTOPPABLE-DOMAINS
 * @example
 * ```ts
 * import { namehash } from "thirdweb/extensions/unstoppable-domains";
 *
 * const result = await namehash({
 *  contract,
 *  labels: ...,
 * });
 *
 * ```
 */
async function namehash(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.labels],
    });
}
