// thirdweb/dist/cjs/extensions/unstoppable-domains/__generated__/UnstoppableDomains/read/reverseNameOf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FN_SELECTOR = void 0;
exports.isReverseNameOfSupported = isReverseNameOfSupported;
exports.encodeReverseNameOfParams = encodeReverseNameOfParams;
exports.encodeReverseNameOf = encodeReverseNameOf;
exports.decodeReverseNameOfResult = decodeReverseNameOfResult;
exports.reverseNameOf = reverseNameOf;
const read_contract_js_1 = require("../../../../../transaction/read-contract.js");
const encodeAbiParameters_js_1 = require("../../../../../utils/abi/encodeAbiParameters.js");
const viem_1 = require("viem");
const detectExtension_js_1 = require("../../../../../utils/bytecode/detectExtension.js");
exports.FN_SELECTOR = "0xbebec6b4";
const FN_INPUTS = [
    {
        type: "address",
        name: "addr",
    },
];
const FN_OUTPUTS = [
    {
        type: "string",
        name: "reverseUri",
    },
];
/**
 * Checks if the `reverseNameOf` method is supported by the given contract.
 * @param availableSelectors An array of 4byte function selectors of the contract. You can get this in various ways, such as using "whatsabi" or if you have the ABI of the contract available you can use it to generate the selectors.
 * @returns A boolean indicating if the `reverseNameOf` method is supported.
 * @extension UNSTOPPABLE-DOMAINS
 * @example
 * ```ts
 * import { isReverseNameOfSupported } from "thirdweb/extensions/unstoppable-domains";
 * const supported = isReverseNameOfSupported(["0x..."]);
 * ```
 */
function isReverseNameOfSupported(availableSelectors) {
    return (0, detectExtension_js_1.detectMethod)({
        availableSelectors,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
    });
}
/**
 * Encodes the parameters for the "reverseNameOf" function.
 * @param options - The options for the reverseNameOf function.
 * @returns The encoded ABI parameters.
 * @extension UNSTOPPABLE-DOMAINS
 * @example
 * ```ts
 * import { encodeReverseNameOfParams } from "thirdweb/extensions/unstoppable-domains";
 * const result = encodeReverseNameOfParams({
 *  addr: ...,
 * });
 * ```
 */
function encodeReverseNameOfParams(options) {
    return (0, encodeAbiParameters_js_1.encodeAbiParameters)(FN_INPUTS, [options.addr]);
}
/**
 * Encodes the "reverseNameOf" function into a Hex string with its parameters.
 * @param options - The options for the reverseNameOf function.
 * @returns The encoded hexadecimal string.
 * @extension UNSTOPPABLE-DOMAINS
 * @example
 * ```ts
 * import { encodeReverseNameOf } from "thirdweb/extensions/unstoppable-domains";
 * const result = encodeReverseNameOf({
 *  addr: ...,
 * });
 * ```
 */
function encodeReverseNameOf(options) {
    // we do a "manual" concat here to avoid the overhead of the "concatHex" function
    // we can do this because we know the specific formats of the values
    return (exports.FN_SELECTOR +
        encodeReverseNameOfParams(options).slice(2));
}
/**
 * Decodes the result of the reverseNameOf function call.
 * @param result - The hexadecimal result to decode.
 * @returns The decoded result as per the FN_OUTPUTS definition.
 * @extension UNSTOPPABLE-DOMAINS
 * @example
 * ```ts
 * import { decodeReverseNameOfResult } from "thirdweb/extensions/unstoppable-domains";
 * const result = decodeReverseNameOfResultResult("...");
 * ```
 */
function decodeReverseNameOfResult(result) {
    return (0, viem_1.decodeAbiParameters)(FN_OUTPUTS, result)[0];
}
/**
 * Calls the "reverseNameOf" function on the contract.
 * @param options - The options for the reverseNameOf function.
 * @returns The parsed result of the function call.
 * @extension UNSTOPPABLE-DOMAINS
 * @example
 * ```ts
 * import { reverseNameOf } from "thirdweb/extensions/unstoppable-domains";
 *
 * const result = await reverseNameOf({
 *  contract,
 *  addr: ...,
 * });
 *
 * ```
 */
async function reverseNameOf(options) {
    return (0, read_contract_js_1.readContract)({
        contract: options.contract,
        method: [exports.FN_SELECTOR, FN_INPUTS, FN_OUTPUTS],
        params: [options.addr],
    });
}
